### DISCUSSION TITLE
How to use mz_zip_reader_entry_read

### DISCUSSION THREAD

**Discussion by zhuda** [nmoinvaz/minizip#612] [Q&A] [answered]
Hi, From the [sample usage code](https://github.com/zlib-ng/minizip-ng/blob/master/doc/mz_zip_rw.md#mz_zip_reader_entry_read) of **mz_zip_reader_entry_read**  it provides a do while loop. This can lead multiple steps to finish one entry reading if the input buffer is too small or the entry is too large. Will this be a valid performance concern if call this API multiple times?

One way I worked around of the usage is leverage the `mz_zip_reader_entry_get_info` API to get entry uncompressed size and allocate a big enough array for `mz_zip_reader_entry_read` so the entry can be completely read in one call. As uncompressed size is int64_t type, and the max number of `mz_zip_reader_entry_read` returned is int32_t type, there still need extra steps if the uncompressed size is larger than max of int32_t;
```
    mz_zip_file* file_info{ nullptr };
    if (mz_zip_reader_entry_get_info(m_zipReader, &file_info) != MZ_OK)
    {
        printf("Could not read file info %s\n", fileName.Value());
        return nullptr;
    }
    std::cout << "entry compress size" << file_info->compressed_size << std::endl;
    std::cout << "entry uncompress size" << file_info->uncompressed_size << std::endl;

    if (mz_zip_reader_entry_open(m_zipReader) == MZ_OK)
    {
        char* buf = new char[file_info->uncompressed_size];
        auto bytes_read = mz_zip_reader_entry_read(m_zipReader, buf, file_info->uncompressed_size);
        if (bytes_read < 0)
        {
            std:cout << "mz_zip_reader_entry_read return invalid code " << bytes_read;
            delete[] buf;
            return nullptr;
        }

        mz_zip_reader_entry_close(m_zipReader);
        delete []buf;
    }
```

I would like to hear if there is any recommended usage of this method.

**Comment by nmoinvaz** [comment, marked as answer]
It can be called a single time if you have enough space for the uncompressed data. Or it can be called multiple times if you do not have a buffer big enough. Calling it once is going to lead to better performance. However, many times it is not possible to create a buffer the size of the uncompressed data, especially if there is a lot of it. If you know you are only going to be dealing with small entries, it makes sense. But if you don't know the size of the entries you are dealing with then it makes sense to just use a good-sized buffer and call it multiple times.

**Reply by zhuda** [reply]
Thanks for the consideration. It makes sense to me.
